transaction/core/src/ring_signature/rct_bulletproofs.rs

```diff
@@ -308,35 +446,110 @@ fn sign_with_balance_check<CSPRNG: RngCore + CryptoRng>(
 
-    let (range_proof, commitments) = {
+    // Range proof is present when mixed transactions are not supported, the set of
+    // range proofs is present when they are.
+    let (range_proof, range_proofs) = if !block_version.mixed_transactions_are_supported() {
         // The implicit fee output is omitted from the range proof because it is known.
+        let generator = generator_cache.get(fee_token_id);
 
         let (values, blindings): (Vec<_>, Vec<_>) = pseudo_output_values_and_blindings
             .iter()
-            .chain(output_values_and_blindings.iter())
-            .map(|(value, blinding)| (*value, *blinding))
+            .cloned()
+            .chain(
+                output_secrets
+                    .iter()
+                    .map(|secret| (secret.value, secret.blinding)),
+            )
             .unzip();
-        generate_range_proofs(&values, &blindings, &generator, rng)
-            .map_err(|_e| Error::RangeProofError)?
+        let (range_proof, _commitments) =
+            generate_range_proofs(&values, &blindings, generator, rng)?;
+
+        (range_proof.to_bytes().to_vec(), vec![])
+    } else {
+        let mut range_proofs = Vec::default();
+
+        // Collect list of of unique token ids
+        let token_ids = {
+            let mut token_ids = BTreeSet::default();
+            token_ids.insert(fee_token_id);
+            for secret in input_secrets {
+                token_ids.insert(secret.token_id);
+            }
+            for secret in output_secrets {
+                token_ids.insert(secret.token_id);
+            }
+            token_ids
+        };
+
+        for token_id in token_ids {
+            let generator = generator_cache.get(token_id);
+
+            // The input blinding is not the same as corresponding pseudo-output blinding
+            let (values, blindings): (Vec<_>, Vec<_>) = input_secrets
+                .iter()
+                .zip(pseudo_output_blindings.iter())
+                .filter_map(|(secret, blinding)| {
+                    if secret.token_id == token_id {
+                        Some((secret.value, *blinding))
+                    } else {
+                        None
+                    }
+                })
+                .chain(output_secrets.iter().filter_map(|secret| {
+                    if secret.token_id == token_id {
+                        Some((secret.value, secret.blinding))
+                    } else {
+                        None
+                    }
+                }))
+                .unzip();
+
+            assert!(
+                !values.is_empty(),
+                "logic error when accumulating commitments for token id"
+            );
+
+            let (range_proof, _commitments) =
+                generate_range_proofs(&values, &blindings, generator, rng)?;
+
+            range_proofs.push(range_proof.to_bytes());
+        }
+
+        (Default::default(), range_proofs)
```

Comment:
Nit: Above you used `vec![]` instead of `Default::default()`, I think that is easier to read/understand.